File: src/engine/gift-mapper.ts
import type { EventBus, GiftEvent, StrengthChangeEvent } from "./event-bus"
import type { ConfigStore } from "../config/store"
import type { GiftRule } from "../config/types"

export class GiftMapper {
  private eventBus: EventBus
  private config: ConfigStore

  constructor(config: ConfigStore, eventBus: EventBus) {
    this.config = config
    this.eventBus = eventBus
    this.eventBus.on("gift", (gift) => this.onGift(gift))
  }

  private onGift(gift: GiftEvent): void {
    const rules = this.config.rules
    // first-match: 找到第一条匹配规则就停止，避免同一礼物被多条规则重复加强度
    for (const rule of rules) {
      if (this.matchRule(rule, gift)) {
        this.applyRule(rule, gift)
        return
      }
    }
  }

  private matchRule(rule: GiftRule, gift: GiftEvent): boolean {
    if (rule.giftId && gift.giftId !== rule.giftId) return false
    if (rule.giftName && gift.giftName !== rule.giftName) return false
    if (rule.coinType !== "all" && rule.coinType !== gift.coinType) return false
    return true
  }

  private applyRule(rule: GiftRule, gift: GiftEvent): void {
    const channels: ("A" | "B")[] = rule.channel === "both" ? ["A", "B"] : [rule.channel]
    const delta = rule.strengthAdd * gift.num

    for (const ch of channels) {
      const event: StrengthChangeEvent = {
        channel: ch,
        delta,
        source: "gift",
        giftName: gift.giftName,
        uname: gift.uname,
        duration: rule.duration,
      }
      this.eventBus.emit("strength:change", event)
    }

    this.eventBus.emit("gift:log", {
      ...gift,
      strengthDelta: channels.map(ch => `${ch}+${delta}`).join(" "),
    })
  }
}

File: src/engine/gift_mapper.rs
use crate::config::types::{Channel, GiftEvent, GiftRule, RuleChannel};
use crate::engine::types::GiftLogEvent;

#[derive(Debug, Clone)]
pub(super) struct RuleStrengthDelta {
    pub(super) channel: Channel,
    pub(super) delta: i32,
    pub(super) duration: u64,
}

pub fn match_rule(rule: &GiftRule, gift: &GiftEvent) -> bool {
    if let Some(rule_gift_id) = rule.gift_id
        && gift.gift_id != rule_gift_id
    {
        return false;
    }
    if gift.gift_name != rule.gift_name {
        return false;
    }
    if !rule.coin_type.matches_gift(gift.coin_type) {
        return false;
    }
    true
}

pub(super) fn apply_rule(rule: &GiftRule, gift: &GiftEvent) -> (Vec<RuleStrengthDelta>, String) {
    let channels = rule_channels(rule.channel);
    let raw_delta = (rule.strength_add as u64).saturating_mul(gift.num as u64);
    let delta_i32 = i32::try_from(raw_delta).unwrap_or(i32::MAX);

    let events: Vec<RuleStrengthDelta> = channels
        .iter()
        .filter(|_| raw_delta > 0)
        .map(|&ch| RuleStrengthDelta {
            channel: ch,
            delta: delta_i32,
            duration: rule.duration,
        })
        .collect();

    let mut effects = Vec::new();
    if raw_delta > 0 {
        effects.extend(channels.iter().map(|ch| format!("{ch:?}+{raw_delta}")));
    }
    if let Some(waveform) = rule.waveform.as_ref() {
        effects.push(format!("波形:{waveform}"));
    }

    (events, effects.join(" "))
}

pub fn rule_channels(channel: RuleChannel) -> Vec<Channel> {
    match channel {
        RuleChannel::Both => vec![Channel::A, Channel::B],
        RuleChannel::A => vec![Channel::A],
        RuleChannel::B => vec![Channel::B],
    }
}

pub fn build_gift_log(gift: &GiftEvent, strength_delta: String) -> GiftLogEvent {
    GiftLogEvent {
        gift_id: gift.gift_id,
        gift_name: gift.gift_name.clone(),
        coin_type: gift.coin_type,
        total_coin: gift.total_coin,
        num: gift.num,
        uid: gift.uid,
        uname: gift.uname.clone(),
        timestamp: gift.timestamp,
        strength_delta,
    }
}
